File: backend/pkg/sources/clients/smart/ssrf.go
package smart

import (
	"fmt"
	"net"
	"net/http"
	"net/url"
	"strings"
	"syscall"
	"time"
)

// validateBaseURL guards the user-supplied FHIR base URL before the backend makes any server-side
// request to it. Because the backend fetches from whatever base a source registers, an unvalidated
// base is a Server-Side Request Forgery (SSRF) vector: it could be aimed at the cloud metadata
// endpoint (169.254.169.254) or an internal RFC1918 service. This is defense-in-depth — in the
// single-user self-hosted model the user controls their own base, but the backend runs with network
// reach the user may not have (k8s, the relay), so we refuse obviously-internal targets.
//
// It accepts only http/https URLs that have a host, and rejects IP-literal hosts in the loopback,
// private, link-local, unique-local, or unspecified ranges (plus the well-known cloud metadata
// addresses) and localhost-ish names. On success it returns the base trimmed of any trailing slash.
//
// It is a PRE-CHECK, not the security boundary. It runs before any connection, so it can give a
// clear error at the moment a source is added — but it cannot be relied on, because it inspects a
// string:
//
//   - net.ParseIP accepts only dotted-quad and IPv6 literals, so "2130706433", "0x7f.0.0.1" and
//     "127.1" all parse as nil and skip the IP check entirely, while the resolver reads them as
//     127.0.0.1 (yourphr#484)
//   - a public name can resolve to a private address, either always or only on the second lookup
//     (DNS rebinding)
//   - a redirect is never seen here at all
//
// The boundary is guardedDialer, which judges the RESOLVED address of every connection at dial
// time. Full egress filtering still belongs at the network layer; these two are defence in depth.
//
// AllowInternalHostsForTest, when true, disables the internal-host SSRF guard process-wide,
// regardless of Config.AllowInternalHosts. It exists ONLY so a consumer's test suite can drive the
// full connect+sync flow (including factory-built clients in background jobs) against httptest
// loopback servers. NEVER set it in production. Per-instance loopback in this package's own tests uses
// Config.AllowInternalHosts instead.
var AllowInternalHostsForTest bool

// allowInternal bypasses the internal-host checks (still validating scheme + host). It is wired to
// Config.AllowInternalHosts (per-instance) and the AllowInternalHostsForTest global — both test-only,
// for httptest loopback servers; never set in production.
func validateBaseURL(raw string, allowInternal bool) (string, error) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return "", fmt.Errorf("FHIR base URL is empty")
	}
	u, err := url.Parse(raw)
	if err != nil {
		return "", fmt.Errorf("FHIR base URL is not a valid URL: %w", err)
	}
	if u.Scheme != "http" && u.Scheme != "https" {
		return "", fmt.Errorf("FHIR base URL must be http(s), got %q", u.Scheme)
	}
	host := u.Hostname()
	if host == "" {
		return "", fmt.Errorf("FHIR base URL has no host")
	}
	if !allowInternal && !AllowInternalHostsForTest {
		if isBlockedHostname(host) {
			return "", fmt.Errorf("FHIR base URL host %q is not allowed (internal/loopback)", host)
		}
		if ip := net.ParseIP(host); ip != nil && isBlockedIP(ip) {
			return "", fmt.Errorf("FHIR base URL host %q is a disallowed internal address", host)
		}
	}
	return strings.TrimRight(raw, "/"), nil
}

// isBlockedHostname rejects localhost and the conventional internal-only TLD suffixes.
func isBlockedHostname(host string) bool {
	h := strings.ToLower(strings.TrimSuffix(host, "."))
	if h == "localhost" {
		return true
	}
	for _, suffix := range []string{".localhost", ".local", ".internal"} {
		if strings.HasSuffix(h, suffix) {
			return true
		}
	}
	return false
}

// isBlockedIP rejects IP literals that point inward: loopback, RFC1918/ULA private, link-local
// (which already covers 169.254.0.0/16), and the unspecified address, plus the explicit cloud
// metadata addresses for clarity.
func isBlockedIP(ip net.IP) bool {
	if ip.IsLoopback() || ip.IsPrivate() || ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast() || ip.IsUnspecified() {
		return true
	}
	if ip.Equal(net.ParseIP("169.254.169.254")) || ip.Equal(net.ParseIP("fd00:ec2::254")) {
		return true
	}
	return false
}

// safeBaseURL validates c.FHIRBaseURL and returns the trimmed base used to build every outbound
// request. All request builders go through this so the SSRF guard cannot be bypassed.
func (c Config) safeBaseURL() (string, error) {
	return validateBaseURL(c.FHIRBaseURL, c.AllowInternalHosts)
}

// guardedDialer refuses connections to internal addresses at DIAL time, after resolution.
//
// This is the actual SSRF boundary. validateBaseURL is a pre-check that produces a friendly error;
// it cannot be the boundary, for two reasons that no amount of string inspection fixes:
//
//  1. Go's net.ParseIP accepts only dotted-quad and IPv6 literals, so every other numeric form
//     returned nil and skipped the check entirely — while the system resolver understood them
//     perfectly well (yourphr#484):
//
//     2130706433           -> 127.0.0.1
//     0x7f.0.0.1           -> 127.0.0.1
//     127.1                -> 127.0.0.1
//     2852039166           -> 169.254.169.254   (cloud metadata)
//     0251.0376.0251.0376  -> 169.254.169.254
//
//  2. DNS rebinding. A name can resolve to a public address when validated and an internal one
//     moments later when dialled. Only a check at connection time sees what was actually reached.
//
// Control runs after the address is resolved and before the socket is connected, and it runs for
// EVERY connection — including redirects, which a base-URL check never sees at all.
func guardedDialer(allowInternal bool) *net.Dialer {
	return &net.Dialer{
		Timeout:   30 * time.Second,
		KeepAlive: 30 * time.Second,
		Control: func(network, address string, _ syscall.RawConn) error {
			if allowInternal || AllowInternalHostsForTest {
				return nil
			}
			host, _, err := net.SplitHostPort(address)
			if err != nil {
				// Control is documented to receive a resolved "ip:port"; anything else is
				// unexpected, so fail closed rather than guess.
				return fmt.Errorf("refusing connection to %q: unrecognised address form", address)
			}
			ip := net.ParseIP(host)
			if ip == nil {
				// Also fail closed: by this point the resolver has run, so a non-IP here means
				// something is wrong rather than something clever.
				return fmt.Errorf("refusing connection to %q: not a resolved IP address", address)
			}
			if isBlockedIP(ip) {
				return fmt.Errorf("refusing connection to internal address %s "+
					"(loopback, private, link-local or cloud metadata)", ip)
			}
			return nil
		},
	}
}

// GuardedTransport returns an http.Transport that cannot connect to an internal address.
//
// Built from http.DefaultTransport's settings rather than a bare &http.Transport{}, so connection
// pooling, HTTP/2 and proxy support behave as they do everywhere else — a hand-rolled transport
// silently loses those and the loss is hard to notice.
func GuardedTransport(allowInternal bool) *http.Transport {
	base, ok := http.DefaultTransport.(*http.Transport)
	if !ok {
		base = &http.Transport{}
	}
	t := base.Clone()
	t.DialContext = guardedDialer(allowInternal).DialContext
	return t
}

File: src/http/ssrf.ts
/**
 * SSRF guarding for outbound provider requests — the piece of the transition with no library
 * behind it (yourphr#539).
 *
 * A self-hosted PHR fetches URLs that a provider, or an attacker who can influence a provider's
 * response, supplies — from inside somebody's home network. The Go backend hardened this over
 * several issues; this is the TypeScript equivalent, written to earn the same protection rather
 * than to look like it.
 *
 * THE BOUNDARY IS AT CONNECT TIME, AFTER DNS RESOLUTION. A URL check cannot be the boundary, for
 * two reasons no amount of string inspection fixes:
 *
 *   1. Numeric host forms. A parser that understands only dotted-quad and IPv6 literals returns
 *      "not an IP" for every other form and skips the check, while the system resolver understands
 *      them perfectly well (yourphr#484):
 *
 *        2130706433           -> 127.0.0.1
 *        0x7f.0.0.1           -> 127.0.0.1
 *        127.1                -> 127.0.0.1
 *        2852039166           -> 169.254.169.254   (cloud metadata)
 *        0251.0376.0251.0376  -> 169.254.169.254
 *
 *   2. DNS rebinding. A name can resolve to a public address when validated and to an internal one
 *      moments later when connected. Only a check at connection time sees what was actually reached.
 *
 * So `validateUrl` below is a courtesy that produces a friendly error early. `guardedLookup` is the
 * control, and it runs for every connection — including each redirect hop, which a base-URL check
 * never sees at all.
 */
import { lookup as dnsLookup, type LookupAddress } from 'node:dns';
import { isIP } from 'node:net';
import { Agent as HttpAgent } from 'node:http';
import { Agent as HttpsAgent } from 'node:https';

/** Message every refusal carries, so callers and tests can recognise the guard's own decision. */
export const REFUSAL = 'refusing to connect to an internal address';

/** No host exempted — the normal state, and what every default argument below means. */
const EMPTY_HOSTS: ReadonlySet<string> = new Set<string>();

/**
 * Hosts an OPERATOR named in configuration, exempted from the internal-address refusal.
 *
 * This exists for one shape of dependency the guard would otherwise make impossible: a service the
 * operator runs and addresses by name on their own network — the language-model endpoint that
 * yourphr#594's chat calls, say, at `http://ollama.lan:11434`. Refusing that is not security, it is
 * the guard being wrong about who chose the address.
 *
 * Three properties keep this from becoming the hole `allowInternal` would be:
 *
 *   - It is a SET OF NAMED HOSTS, not a switch. Exempting one host says nothing about
 *     169.254.169.254, and a provider whose configured URI is attacker-influenced still cannot
 *     reach anything but the host the operator wrote down.
 *   - It is supplied ONCE, when the capability is constructed from configuration, and stripped from
 *     the per-request options (see `RequestOptions` in ./index.ts). A call site cannot widen it.
 *   - It is matched on the HOSTNAME the URL carries, before resolution, so it exempts the name the
 *     operator declared rather than whatever that name currently resolves to. A redirect to a
 *     different internal host is still refused.
 */
export function isAllowedHost(host: string, allowHosts: ReadonlySet<string>): boolean {
  if (allowHosts.size === 0) return false;
  return allowHosts.has(host.toLowerCase().replace(/\.$/, ''));
}

/**
 * Hostnames that never leave the machine, refused by name before any resolution.
 *
 * Mirrors the Go guard: localhost and the suffixes that mDNS, container runtimes and cloud
 * providers use for names that are internal by definition.
 */
export function isBlockedHostname(host: string): boolean {
  const h = host.toLowerCase().replace(/\.$/, '');
  if (h === 'localhost') {
    return true;
  }
  return ['.localhost', '.local', '.internal'].some((suffix) => h.endsWith(suffix));
}

function ipv4ToParts(ip: string): number[] | null {
  const parts = ip.split('.');
  if (parts.length !== 4) {
    return null;
  }
  const nums = parts.map((p) => Number(p));
  return nums.every((n) => Number.isInteger(n) && n >= 0 && n <= 255) ? nums : null;
}

/**
 * Addresses that point inward: loopback, RFC1918 private, carrier-grade NAT, link-local (which
 * covers the 169.254.0.0/16 metadata range), the unspecified address, and IPv6 unique-local.
 *
 * Node has no equivalent of Go's ip.IsPrivate()/IsLoopback(), so the ranges are spelled out. The
 * IPv4-mapped IPv6 form (::ffff:127.0.0.1) is unwrapped first — it is a documented bypass when a
 * guard checks only the textual form.
 */
export function isBlockedIp(address: string): boolean {
  let ip = address.trim().toLowerCase();

  // Strip a zone index (fe80::1%eth0) before parsing.
  const zone = ip.indexOf('%');
  if (zone !== -1) {
    ip = ip.slice(0, zone);
  }

  // IPv4-mapped and IPv4-compatible IPv6 — judge the embedded IPv4, not the wrapper.
  const mapped = /^::(?:ffff:)?(\d+\.\d+\.\d+\.\d+)$/.exec(ip);
  if (mapped?.[1]) {
    ip = mapped[1];
  }

  const v4 = ipv4ToParts(ip);
  if (v4) {
    const [a, b] = v4 as [number, number, number, number];
    if (a === 0) return true; // 0.0.0.0/8, includes the unspecified address
    if (a === 10) return true; // RFC1918
    if (a === 127) return true; // loopback
    if (a === 100 && b >= 64 && b <= 127) return true; // RFC6598 carrier-grade NAT
    if (a === 169 && b === 254) return true; // link-local, includes 169.254.169.254 metadata
    if (a === 172 && b >= 16 && b <= 31) return true; // RFC1918
    if (a === 192 && b === 168) return true; // RFC1918
    if (a >= 224) return true; // multicast and reserved
    return false;
  }

  if (isIP(ip) !== 6) {
    // Not a form this function can judge. Refuse rather than allow: an address we cannot parse is
    // not an address we can vouch for, and the caller's alternative is connecting blind.
    return true;
  }

  if (ip === '::' || ip === '::1') return true; // unspecified, loopback
  if (/^f[cd]/.test(ip)) return true; // fc00::/7 unique-local, includes fd00:ec2::254 metadata
  if (/^fe[89ab]/.test(ip)) return true; // fe80::/10 link-local
  if (ip.startsWith('ff')) return true; // multicast
  return false;
}

/**
 * A DNS lookup that refuses to hand back an internal address.
 *
 * Passed as the `lookup` option to an http.Agent, so Node calls it for every connection the agent
 * makes. Resolution happens first and every returned address is judged — `all: true` matters,
 * because a name with several A records must not pass on the strength of one public answer while
 * another points inward.
 */
export function guardedLookup(allowInternal = false, allowHosts: ReadonlySet<string> = EMPTY_HOSTS): typeof dnsLookup {
  const guarded = (hostname: string, options: unknown, callback: unknown): void => {
    // Node calls lookup(hostname, options, cb) or lookup(hostname, cb).
    const cb = (typeof options === 'function' ? options : callback) as (
      err: NodeJS.ErrnoException | null,
      address?: string | LookupAddress[],
      family?: number
    ) => void;
    const opts = (typeof options === 'function' ? {} : options ?? {}) as Record<string, unknown>;

    const exempt = allowInternal || isAllowedHost(hostname, allowHosts);

    if (!exempt && isBlockedHostname(hostname)) {
      cb(Object.assign(new Error(`${REFUSAL}: ${hostname}`), { code: 'ESSRFBLOCKED' }));
      return;
    }

    dnsLookup(hostname, { ...opts, all: true }, (err, addresses) => {
      if (err) {
        cb(err);
        return;
      }
      const resolved = addresses as LookupAddress[];
      if (!exempt) {
        const blocked = resolved.find((a) => isBlockedIp(a.address));
        if (blocked) {
          cb(
            Object.assign(new Error(`${REFUSAL}: ${hostname} resolved to ${blocked.address}`), {
              code: 'ESSRFBLOCKED',
            })
          );
          return;
        }
      }
      if (opts['all'] === true) {
        cb(null, resolved);
      } else {
        const first = resolved[0];
        if (!first) {
          cb(Object.assign(new Error(`no addresses for ${hostname}`), { code: 'ENOTFOUND' }));
          return;
        }
        cb(null, first.address, first.family);
      }
    });
  };
  return guarded as unknown as typeof dnsLookup;
}

export interface GuardedAgents {
  http: HttpAgent;
  https: HttpsAgent;
}

/** Agents whose every connection goes through the guarded lookup. */
export function guardedAgents(allowInternal = false, allowHosts: ReadonlySet<string> = EMPTY_HOSTS): GuardedAgents {
  const options = { lookup: guardedLookup(allowInternal, allowHosts), keepAlive: false };
  return { http: new HttpAgent(options), https: new HttpsAgent(options) };
}

/**
 * An early, friendly check on a base URL. NOT the boundary — see the header. It exists so an
 * operator typing a bad address gets a clear message instead of a connection error, and it
 * deliberately refuses an unparseable URL rather than passing it along.
 */
export function validateUrl(raw: string, allowInternal = false, allowHosts: ReadonlySet<string> = EMPTY_HOSTS): { ok: true; url: URL } | { ok: false; reason: string } {
  let url: URL;
  try {
    url = new URL(raw);
  } catch {
    return { ok: false, reason: 'not a valid URL' };
  }
  if (url.protocol !== 'https:' && url.protocol !== 'http:') {
    return { ok: false, reason: `unsupported scheme ${url.protocol}` };
  }
  if (allowInternal) {
    return { ok: true, url };
  }
  // URL puts IPv6 literals in brackets.
  const host = url.hostname.replace(/^\[|\]$/g, '');
  if (isAllowedHost(host, allowHosts)) {
    return { ok: true, url };
  }
  if (isBlockedHostname(host)) {
    return { ok: false, reason: `${REFUSAL}: ${host}` };
  }
  if (isIP(host) !== 0 && isBlockedIp(host)) {
    return { ok: false, reason: `${REFUSAL}: ${host}` };
  }
  return { ok: true, url };
}
